### ISSUE TITLE
[Proposal] Android-Specific Navigation Bar Effect

### ISSUE THREAD

**Issue by brminnick** [CommunityToolkit/Maui#103]
# NavigationBarEffect

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
  * [ ] iOS Support
  * [ ] Android Support
  * [ ] macOS Support
  * [ ] Windows Support
* [ ] Unit Tests: Not Started
* [ ] Sample: Not Started
* [ ] Documentation: Not Started

## Summary
[summary]: #summary

Android-specific implementation allowing the user to customize the Navigation Bar

## Detailed Design
[design]: #detailed-design

### NavigationBarEffect.cs

```cs
namespace Xamarin.CommunityToolkit.PlatformConfiguration.AndroidSpecific
{
	[Preserve(Conditional = true)]
	public static class NavigationBarEffect
	{
		static NavigationBarEffect()
		{
			#region Required work-around to prevent linker from removing the platform-specific implementation
#if __ANDROID__
			if (System.DateTime.Now.Ticks < 0)
				_ = new PlatformNavigationBarEffect();
#endif
			#endregion
		}

		public static readonly BindableProperty ColorProperty = BindableProperty.CreateAttached(
			"Color", typeof(Color), typeof(NavigationBarEffect), Color.Default, propertyChanged: TryGenerateEffect);

		public static readonly BindableProperty StyleProperty = BindableProperty.CreateAttached(
			"Style", typeof(NavigationBarStyle), typeof(NavigationBarEffect), NavigationBarStyle.Default, propertyChanged: TryGenerateEffect);

		public static Color GetColor(BindableObject bindable) =>
			(Color)bindable.GetValue(ColorProperty);

		public static void SetColor(BindableObject bindable, Color value) =>
			bindable.SetValue(ColorProperty, value);

		public static NavigationBarStyle GetStyle(BindableObject bindable) =>
			(NavigationBarStyle)bindable.GetValue(StyleProperty);

		public static void SetStyle(BindableObject bindable, NavigationBarStyle value) =>
			bindable.SetValue(StyleProperty, value);

		public static IPlatformElementConfiguration<XFP.Android, FormsElement> SetNavigationBarColor(this IPlatformElementConfiguration<XFP.Android, FormsElement> config, Color color)
		{
			SetColor(config.Element, color);
			return config;
		}

		public static Color GetNavigationBarColor(this IPlatformElementConfiguration<XFP.Android, FormsElement> config) =>
			GetColor(config.Element);

		public static IPlatformElementConfiguration<XFP.Android, FormsElement> SetNavigationBarStyle(this IPlatformElementConfiguration<XFP.Android, FormsElement> config, NavigationBarStyle style)
		{
			SetStyle(config.Element, style);
			return config;
		}

		public static NavigationBarStyle GetNavigationBarStyle(this IPlatformElementConfiguration<XFP.Android, FormsElement> config) =>
			GetStyle(config.Element);

		static void TryGenerateEffect(BindableObject bindable, object oldValue, object newValue)
		{
			if (bindable is not FormsElement page)
				return;

			DetachEffect(page);

			AttachEffect(page);
		}

		static void AttachEffect(FormsElement element)
		{
			IElementController controller = element;
			if (controller is null || controller.EffectIsAttached(EffectIds.NavigationBar))
				return;

			element.Effects.Add(Effect.Resolve(EffectIds.NavigationBar));
		}

		static void DetachEffect(FormsElement element)
		{
			IElementController controller = element;
			if (controller is null || !controller.EffectIsAttached(EffectIds.NavigationBar))
				return;

			var toRemove = element.Effects.FirstOrDefault(e => e.ResolveId == Effect.Resolve(EffectIds.NavigationBar).ResolveId);
			if (toRemove != null)
				element.Effects.Remove(toRemove);
		}
	}
}
```

## Usage Syntax
[usage]: #usage-syntax

### XAML Usage

```xml
<Setter Property="droid:NavigationBarEffect.Color" Value="{StaticResource NavBarColor}" />
```

### C# Usage

```cs
var navigationPage = new NavigationPage();
NavigationBarEffect.SetColor(navigationPage, Colors.Blue);
```

**Comment by brminnick** [comment]
On hold, pending results of https://github.com/dotnet/maui/pull/2049

**Comment by maxkoshevoi** [comment]
Would like to implement this, in case issue is unblocked

**Comment by maxkoshevoi** [comment]
dotnet/maui#2049 is now closed, can we move forward with thins issue?

**Comment by pictos** [comment]
This is still blocked by #124

**Comment by maxkoshevoi** [comment]
Since it seems like https://github.com/CommunityToolkit/Maui/issues/85#issuecomment-1239441230 is getting some traction, could this one be moved forward as well?

**Comment by pictos** [comment]
@maxkoshevoi we will vote to champion this after the Status bar implementation

**Comment by tranb3r** [comment]
Any update? Thanks

**Comment by tranb3r** [comment]
@pictos @brminnick 
What is exactly the status of this issue? Thanks

**Comment by brminnick** [comment]
I don't have any additional information for you than what is already provided.

On the [Project Board](https://github.com/CommunityToolkit/Maui/projects/1) this Proposal is in the `Project Submitted` column. 

It currently has the https://github.com/CommunityToolkit/Maui/labels/new label. 

As noted in our [`Submitting a New Feature` guidelines](https://github.com/CommunityToolkit/Maui?tab=readme-ov-file#submitting-a-new-feature), the `Proposal Submitted` column indicates the following: 
> Once you have a fully fleshed out proposal describing a new feature in syntactic and semantic detail, please [open an issue for it](https://github.com/communitytoolkit/maui/issues/new/choose), and it will be labeled as a [Proposal](https://github.com/communitytoolkit/maui/issues?q=is%3Aopen+is%3Aissue+label%3Aproposal). The comment thread on the issue can be used to hash out or briefly discuss details of the proposal, as well as pros and cons of adopting it into the .NET MAUI Toolkit. If an issue does not meet the bar of being a full proposal, we may move it to a discussion, so that it can be further matured. Specific open issues or more expansive discussion with a proposal will often warrant opening a side discussion rather than cluttering the comment section on the issue.

It is not actively being worked by any member of the community.

**Comment by brminnick** [comment]
I am inclined to close this Proposal as it overlaps with our existing feature, [StatusBarBehavior](https://learn.microsoft.com/dotnet/communitytoolkit/maui/behaviors/statusbar-behavior?tabs=ios).

**Comment by maxkoshevoi** [comment]
> I am inclined to close this Proposal as it overlaps with our existing feature, [StatusBarBehavior](https://learn.microsoft.com/dotnet/communitytoolkit/maui/behaviors/statusbar-behavior?tabs=ios).
> 

This is different though. Status bar is at the top while the navigation bar is at the bottom (where navigation buttons are).
Or do you mean that this functionality can be included in the StatusBarBehavior somehow?

**Comment by brminnick** [comment]
> This is different though. Status bar is at the top while the navigation bar is at the bottom (where navigation buttons are).

Thank you for clarifying. I will keep this Proposal open.

**Comment by brminnick** [comment]
Thanks Max! I will present this Proposal to the team at our Monthly Standup on Friday: https://www.youtube.com/watch?v=a6LBam_9q-0

Assuming at least 50% of the maintainers approve, we'll move to it Approved and I'll assign it to you 👍

**Comment by jfversluis** [comment]
Alright let's get this party started! I think we have all the ingredients to get this going. Thanks for taking this up @maxkoshevoi.

A few things:
* We're going to implement this as a Behavior because effects are not a thing anymore in .NET MAUI. For inspiration look at #617
* This is specific to only Android, I'm not sure if we have anything like that already in the .NET MAUI Toolkit?
* The name is a bit confusing, Brandon thought it was about the navigation bar in the top of the screen. Is there any better name that we can come up with maybe? Is navigation bar really the name that Android has for that bar as well?
* Other than that I think it would be good to keep the API as close as the XCT one as much as possible so transitioning will be easier.

Let me know if there are any issues or questions!

**Comment by tranb3r** [comment]
Names of system bars on android:
- status bar: https://developer.android.com/design/ui/mobile/guides/foundations/system-bars#status-bar
- navigation bar: https://developer.android.com/design/ui/mobile/guides/foundations/system-bars#navigation-bar

**Comment by bijington** [comment]
I do wonder if we should have named `StatusBarBehavior` as `SystemBarBehavior` and then have properties for the 2 concepts. Possibly a bit late now though

**Comment by tranb3r** [comment]
Anybody working on this?
About naming, what is your decision?

**Comment by jfversluis** [comment]
@tranb3r @maxkoshevoi expressed interest but I don't think much has happened since then. So as far as I am concerned you're free to take this up.

For naming... BottomNavigationBar so that it's clear its that bottom one? SystemNavigationBar? I'm open to suggestions.

**Comment by maxkoshevoi** [comment]
@tranb3r , yes sorry, got unexpectedly busy. You can take this

**Comment by tranb3r** [comment]
No problem. I'll try to do it today. Should not be very long, since it's basically copy/paste from StatusBar.
About naming: I think SystemNavigationBar is better. It's how android is naming it in the docs. And it avoids confusion with any application navbar (top, bottom, tabs, whatever).

**Comment by jfversluis** [comment]
Sounds good to me!

**Comment by tranb3r** [comment]
@pictos 
I can move both properties of the behavior to android specific.
But the behavior itself has to be cross-platform, right?

Is this the usage you want?
```
<Page.Behaviors>
    <OnPlatform x:TypeArguments="Behavior">
        <On Platform="Android">
            <mct:SystemNavigationBarBehavior 
                android:SystemNavigationBarBehavior.SystemNavigationBarColor="{Binding ...}" 
                android:SystemNavigationBarBehavior.SystemNavigationBarStyle="{Binding ...}" />
        </On>
    </OnPlatform>
</Page.Behaviors>
```
```
On<Android>().SetSystemNavigationBarColor(Colors.Red);
```

**Comment by pictos** [comment]
@tranb3r that shouldn't be Behaviors, but AttachedProperties

**Comment by pictos** [comment]
@tranb3r sorry, for the confusion... The OnPlatform API was implemented as Effect, like all `OnPlatform` stuff, on XCT, but on .NET MAUI that concept is _obsolete_ and we moved all Effects for the [PlatformBehavior API](https://github.com/dotnet/maui/issues/4958). And this is the first feature like that we will implement here, so we're learning in the path.

Talking about this feature, it's a way to expose a property for a single platform in a shared layer, so in the end, isn't the same, and the `PlatformBehavior` could not fit very well, because it will hide from users that Behavior will just work for a specific platform... 

As I mentioned in the PR, the best solution is to follow how .NET MAUI implements it. I know that could be frustrating, to spend time and be asked to change a go amount of it, so if you want, I can jump in and finish the implementation. I'll do above your PR so the credit is still yours, for taking time to help us

**Comment by tranb3r** [comment]
@pictos I understand. No problem.
I'm just looking forward for this _simple_ feature to be implemented (and I really do not care about the credit).
So please go ahead if you have the time to work on it. It's not a big change anyway.
Thanks!

**Comment by pictos** [comment]
@tranb3r I can jump in or guide you, I'll let you choose what's more fun for you.

> For me, since you're doing OSS at your free time, you should choose what is more fun for you

**Comment by tranb3r** [comment]
@pictos 
Ok, I'm happy to do it, although there is something I do not understand.
Let's assume we implement two AttachedProperties (`SystemNavigationBarColor` and `SystemNavigationBarStyle`). 
I guess they would be attached to the `Page`, right?
I easily understand how to create the properties, in android-specific platformConfiguration. This would be similar to what has been done in dotnet maui for UseSafeArea.
However this is not enough. The code needs to react to the properties change. For StatusBarColor this is done in the Behavior, but we said we do not want to use a Behavior for SystemNavigationBar. For UseSafeArea in dotnet maui, this is done in the `PageRenderer` itself. Where should we put this code in our case?

**Comment by pictos** [comment]
> @pictos Ok, I'm happy to do it, although there is something I do not understand. Let's assume we implement two AttachedProperties (`SystemNavigationBarColor` and `SystemNavigationBarStyle`). I guess they would be attached to the `Page`, right? I easily understand how to create the properties, in android-specific platformConfiguration. This would be similar to what has been done in dotnet maui for UseSafeArea. However this is not enough. The code needs to react to the properties change. For StatusBarColor this is done in the Behavior, but we said we do not want to use a Behavior for SystemNavigationBar. For UseSafeArea in dotnet maui, this is done in the `PageRenderer` itself. Where should we put this code in our case?

TBH I don't know, we need to test a couple of approaches, but my first attempt will be to use the Handlers mappings to attach the behavior that we want for on the propertiesChanges that we want... Another possibility, (that I don't like too much) is to use the `PropertyChanged` on the attached property to listen for the changes that will happen on `Page`

**Comment by tranb3r** [comment]
To be honest, the android API for this feature is _extremely_ simple, so it's really frustrating to be confronted to so much complexity for implementing it in maui. 
I mean, if even you Pedro do not know how to do it without experimenting different approaches, I'm probably wasting my time here.
So, I'm really sorry, but I won't work on it anymore. I hope you can understand my decision. And I'll be thankful to you if you finish the implementation!

**Comment by jfversluis** [comment]
@tranb3r welcome to our world! 😄 APIs might be very simple on the platform itself but we do want to make sure that it's surfaced to .NET MAUI in a way that makes sense for that ecosystem. The way we leveraged before in Xamarin.Forms does not exist anymore in .NET MAUI so we have to find something new.

Sorry to hear that frustrates you, but if we are to accept any feature, big or small, we want to be able to be convinced about it as we are going to be the ones responsible for maintaining it and not go around and add breaking changes in every other version because we changed our minds.

**Comment by tranb3r** [comment]
@jfversluis I totally understand and I'd really like to support this effort, just like I contributed to XCT back in the days. 

But here, my feeling is that if an API is simple on the platform itself, it sould be simple to surface it to Maui, without too much complexity in the implementation. 
Ideally, it should be as simple as implementing a SetSystemNavigationBarColor method in a .android.cs file. And Maui would take care of all the plumbing.
But it's not at all the case. And if even you guys cannot figure out how to do it, it's probably because something is missing or too complex or maybe wrong in the framework.

Again, I'm 100% in favor of doing the proper implementation. For all the good reasons you mentionned, and also because I like to do things properly. It's just that so far, I do not understand how this is supposed to be implemented. I'll be happy to read the code when you figure out what's the good implementation.

**Comment by pictos** [comment]
@tranb3r, this project wants to make its implementation as close to the .NET MAUI as possible, that's because we want:

1. Make it easier for people who try to use the features, they will have to think the same way they think when using MAUI, so they don't need to learn a new way to do things;
2. We always try to promote some features from here to the MAUI framework, so implementing as closest they do makes it easier to port the feature

Even if the API is simple and should be just one line, we need to think about a lot of stuff, like "Will there be more features like that?",  "Is this API cross-platform friendly? (we want to avoid users having to type #if #endif to use the API)", etc. Including also what @jfversluis said.

You probably know, but is good to remember that the mindset of developing an application is very different from the mindset of building a library, we need to be more careful, and for a lib with visibility like ours we need to be even more careful when implementing new APIs, for context, before move way from Effects we have had a lot of PoCs, discussions and tests to make sure that's the correct path and at the end I opened a proposal on .NET MAUI for PlatformBehavior and did the implementation (with the help of the team).

Talking about this proposal, I'll repeat myself, .NET MAUI already has a way to handle Platform-Specific APIs, that's APIs that work only for one platform and are exposed in the shared layer, we must follow that approach as much as we can, because of the items 1 and 2 that I mentioned above. I hope this clarifies the technical side of decisions we made

**Comment by tranb3r** [comment]
Again I totally agree. No need to repeat yourself or try to justify anything.

But you have to admit that right now, nobody has been able to explain clearly how this feature should be implemented.
This is why I have the feeling that the framework is a bit too complex. Especially when dealing with such a simple feature.
But I insist, this is not a big deal. And I'd really like this discussion to keep being focused on how to implement the feature.

So, let me clarify what I'm expecting here:
- We all know about the platform specific API for properties. The code is simple. The UseSafeArea is a good example of it.
- What is not clear right now is _all the rest_. How do we react to this platform-specific property changed (like it was done with the Behavior)? If you know how to implement it, please just do it. I'll be happy to learn something.

**Comment by jfversluis** [comment]
All done, released as part of [v8.0.0](https://github.com/CommunityToolkit/Maui/releases/tag/8.0.0)
